### ISSUE TITLE
Lab 7 level 4

### ISSUE THREAD

**Issue by mechastriker3** [nus-cs2030/2021-s1#241]
Hi anyone know how to ensure that type T always extends from comparable? I tried doing Lazy<T extends Comparable<T>> and it threw a whole bunch of type argument out of bounds error. 

## Topic: 
Java Generics

**Comment by domlimm** [comment]
Use `Lazy<T extends Comparable<T>>` instead!

**Comment by mechastriker3** [comment]
Oh I tried that it gave the same error. Do you have to change all the parameterized types to <T extends Comparable<T>>

**Comment by Chen-Yiyang** [comment]
> Oh I tried that it gave the same error. Do you have to change all the parameterized types to <T extends Comparable>

Yes I think you should do `<T extends Comparable<T>>` for all the param types, including new instances you return.

**Comment by bryanwee023** [comment]
@mechastriker3 If don't want to change all the types, maybe an easier method would be to just limit the type of the static factory method's argument to `T extends Comparable<T>`. Y'all think this invites any drawbacks?

**Comment by merylseow** [comment]
Yeah I think you should declare it in every method, eg. `public <T extends Comparable<T>> Lazy<T> map(Function<? super T, ? extends U> f)`.

**Comment by weiyifei-evelyn** [comment]
Yes, you should declare T extends Comparable<T> in all the methods using T. Here are some examples:
1. when defining the class
`class Lazy<T extends Comparable<T>>`
2. when defining method of()
`static <T extends Comparable<T>> Lazy<T> of(T v)`
3. when defining other method which returns Lazy<R>
`public <R extends Comparable<R>> Lazy<R> flatMap(Function<? super T, ? extends Lazy<? extends R>> mapper)`

Hope these examples help you!
